File: src/git/repository.ts
// The meal-plan folder is a git repository, and the server keeps it that way.
//
// EVERY git command runs INSIDE THE SANDBOX. That is not a style choice.
//
// The bind mounts the whole folder, `.git` included, so the agent can write
// `.git/hooks/pre-commit`, or a `filter.*.clean` in `.git/config` with a
// matching `.gitattributes`. Git runs both of those as ordinary programs. Had
// the server committed with the host's git, an agent would have had arbitrary
// execution on the host, outside every boundary this project builds — from a
// recipe file. Running git in the sandbox puts hooks and filters back inside
// it, where the image has no interpreter and no network for them to use.
//
// It also means the product needs no git on the host at all.

import type { RunResult, Session } from '../sandbox/session.ts';

/**
 * The identity the server commits under. There is no /etc/passwd in the image,
 * so git has nothing to guess from — which is the point. It is written into the
 * repository's own config so that a `git revert` or `git commit` the AGENT runs
 * has an identity too.
 */
export const COMMITTER = {
  name: 'Meal Planner',
  email: 'meal-planner@localhost',
};

export const FIRST_COMMIT_MESSAGE = 'Initialise the meal plan folder';

export type Clock = () => Date;

/** Git wants RFC 2822 or ISO 8601. ISO with an offset is unambiguous. */
export function gitDate(at: Date): string {
  return at.toISOString().replace('T', ' ').replace(/\.\d+Z$/, ' +0000');
}

export function commitEnvironment(at: Date): Record<string, string> {
  const stamp = gitDate(at);
  return {
    GIT_AUTHOR_NAME: COMMITTER.name,
    GIT_AUTHOR_EMAIL: COMMITTER.email,
    GIT_AUTHOR_DATE: stamp,
    GIT_COMMITTER_NAME: COMMITTER.name,
    GIT_COMMITTER_EMAIL: COMMITTER.email,
    GIT_COMMITTER_DATE: stamp,
  };
}

/**
 * Make the folder a repository with a first commit, if it is not one already.
 *
 * The first commit matters: `git restore --source=HEAD~1` and `git revert` are
 * the undo button the folder otherwise does not have, and neither works in a
 * repository with no history.
 */
export async function ensureRepository(session: Session, now: Clock): Promise<void> {
  const existing = await session.run('git rev-parse --git-dir', { commit: false });
  if (existing.exitCode === 0) return;

  const at = now();
  const result = await session.run(
    [
      'set -e',
      'git init -q -b main',
      `git config user.name ${quote(COMMITTER.name)}`,
      `git config user.email ${quote(COMMITTER.email)}`,
      'git add -A',
      `git commit -q -m ${quote(FIRST_COMMIT_MESSAGE)}`,
    ].join('\n'),
    { commit: false, env: commitEnvironment(at) },
  );

  if (result.exitCode !== 0) {
    throw new Error(
      `could not make ${session.folder} a git repository:\n${result.stderr || result.stdout}`,
    );
  }
}

/** True when the working tree differs from HEAD. */
export async function isDirty(session: Session): Promise<boolean> {
  const status = await session.runDirect('git status --porcelain', { commit: false });
  return status.stdout.trim() !== '';
}

/**
 * Stage everything and make one commit. The caller decides whether there is
 * anything to commit; see src/git/commit.ts.
 */
export async function commitAll(
  session: Session,
  message: string,
  at: Date,
): Promise<RunResult> {
  // The message is the command line the agent ran, so it is agent-controlled
  // text. It travels as an environment variable and is expanded inside double
  // quotes, which bash does not re-parse. A heredoc would have been an
  // injection: a command line holding the delimiter closes it, and everything
  // after it runs.
  return session.runDirect('git add -A && git commit -q -m "$MEALPLAN_COMMIT_MESSAGE"', {
    commit: false,
    env: { ...commitEnvironment(at), MEALPLAN_COMMIT_MESSAGE: message },
  });
}

/** Single-quote for bash. The only safe way to put agent text on a command line. */
export function quote(text: string): string {
  return `'${text.replaceAll("'", `'\\''`)}'`;
}

File: lib/mealplan/git/repository.ex
defmodule Mealplan.Git.Repository do
  @moduledoc """
  Make and inspect the meal-plan folder's git repository. Ported from
  `src/git/repository.ts`. Every git command runs inside the sandbox session.
  """

  alias Mealplan.Git.Commit
  alias Mealplan.Sandbox.Session

  @doc """
  Make the folder a repository with a first commit, if it is not one already.

  The first commit matters: `git restore --source=HEAD~1` and `git revert` are
  the undo button the folder otherwise does not have, and neither works in a
  repository with no history.
  """
  @spec ensure_repository(pid(), DateTime.t()) :: :ok
  def ensure_repository(session, %DateTime{} = at) do
    existing = Session.run(session, "git rev-parse --git-dir")

    if existing.exit_code == 0 do
      :ok
    else
      script =
        Enum.join(
          [
            "set -e",
            "git init -q -b main",
            "git config user.name #{Commit.quote_arg(Commit.committer_name())}",
            "git config user.email #{Commit.quote_arg(Commit.committer_email())}",
            "git add -A",
            "git commit -q -m #{Commit.quote_arg(Commit.first_commit_message())}"
          ],
          "\n"
        )

      result = Session.run(session, script, env: Commit.commit_environment(at))

      if result.exit_code != 0 do
        raise "could not make the meal-plan folder a git repository:\n" <>
                if(result.stderr == "", do: result.stdout, else: result.stderr)
      end

      :ok
    end
  end

  @doc "True when the working tree differs from HEAD."
  @spec dirty?(pid()) :: boolean()
  def dirty?(session) do
    status = Session.run(session, "git status --porcelain")
    String.trim(status.stdout) != ""
  end

  @doc "Number of commits reachable from HEAD, or 0 for a repo with no history."
  @spec commit_count(pid()) :: non_neg_integer()
  def commit_count(session) do
    counted = Session.run(session, "git rev-list --count HEAD")

    case Integer.parse(String.trim(counted.stdout)) do
      {n, _} -> n
      :error -> 0
    end
  end

  @doc """
  Last 3 commit subjects and the first 5 files changed in HEAD, for the
  session-open tree view.
  """
  @spec recent_history(pid()) :: String.t()
  def recent_history(session) do
    log = Session.run(session, "git log --oneline -3 --no-decorate")

    files =
      Session.run(
        session,
        "git diff-tree --no-commit-id --name-only -r HEAD 2>/dev/null | head -5"
      )

    lines = ["recent commits:"]

    lines =
      if log.exit_code == 0 and String.trim(log.stdout) != "" do
        lines ++ Enum.map(String.split(String.trim(log.stdout), "\n"), &("  " <> &1))
      else
        lines
      end

    lines =
      if files.exit_code == 0 and String.trim(files.stdout) != "" do
        lines ++
          ["", "files in HEAD:"] ++
          Enum.map(String.split(String.trim(files.stdout), "\n"), &("  " <> &1))
      else
        lines
      end

    Enum.join(lines, "\n")
  end
end
